### PR TITLE
Switch to using result

### PR DESCRIPTION
**Pull request by rgrinberg**
Switch from polymorphic variant version of result to the standard one.

This is breaking a change but I think there's a manageable number of packages
to fix:
```
ketrew
key-parsers
ojs-base
ppx_json_types
socket-daemon
```
@nv-vn @zoggy @smondet @emillon

### CODE REVIEW

**Comment by whitequark** [comment]
I'm not sure how much benefit does this bring in...

**Comment by rgrinberg** [comment]
Strictly speaking, changing this plugin to use result is much more important to me than the equivalent concern I raised in ppx_deriving itself. There it's just a nice to have that I can live without, but here not being able to use rresult's function (or core/containers) is a nuisance.

I don't mind helping out migrating code that's on opam to get this in if that's the concern. Do you have some internal codebase that will be annoying to migrate?

**Comment by hcarty** [comment]
For what it's worth, I agree with @rgrinberg on this.

**Comment by whitequark** [comment]
OK. I will merge this and do a major version bump once I have time.

**Comment by whitequark** [comment]
One thing I would like to see is the `module List = List` etc removed from Ppx_deriving_yojson_runtime as those should now be handled by hygiene in ppx_deriving. So to minimize the amount of breakage overall, could you please do this:
  * Send a PR to ppx_deriving that adds a dependency on `result`, switches the internal API to use `Result` and, since we already have the dependency, puts `Result` into `Ppx_deriving_runtime`.
  * Extend `Ppx_deriving.sanitize` to accept a module name instead of hardcoding `Ppx_deriving_runtime` (which would be a breaking change otherwise).
  * Migrate ppx_deriving_yojson to use ppx_deriving's hygiene API and use the updated `Ppx_deriving.sanitize` function to use `Ppx_deriving_yojson_runtime`, which would now `include Ppx_deriving_runtime` instead of listing stdlib modules directly.
  * Ensure that all opam dependents of `ppx_deriving` and `ppx_deriving_yojson` have correct version constraints.

**Comment by emillon** [comment]
I am in favor of this change too, since we're consolidating the error conventions in our codebase and this one is the one we have the least control over :)

If that helps, we're depending on the representation mainly in the following ways:

  - use a `of_string` function to make a `of_yojson` function; [example](https://github.com/cryptosense/key-parsers/blob/43631f7b605c3fac4a8b52be9012202bb65a96e4/src/asn1.ml#L16)
  - use a different JSON representation for a type, for example the following module (de)serializes strings using a list of signed chars:

```ocaml
type t = string

let of_yojson j =
  let open Yojson_monad in (* internal module defining (>>=), mapM, error, combinators for field access etc *)
  [%of_yojson: int list] j >>= fun l ->
  `Ok (string_of_signed_byte_list l)

let to_yojson x =
  [%to_yojson: int list] @@ string_to_signed_byte_list x
```

In short, I believe that we only use the monadic interface. So, if there was a `Ppx_deriving_yojson_runtime.return` and we used that instead of the constructor, we could be compatible with both the polymorphic variant version and the `result` one. Would that be an option?

**Comment by whitequark** [comment]
I think that, if there is consensus that using `Result` is the proper way forward, then we should go ahead and break the API. It's just that it wasn't clear to me before that there were definitive advantages. Now I think I do see it.

**Comment by rgrinberg** [comment]
> So to minimize the amount of breakage overall, could you please do this:

This is a good plan and I'm currently following it but I think we can actually minimize some opam constraint hell by deviating a little. Switching ppx_deriving's internal api to use result actually did not break this plugin. Also, it seems to me that `sanitize` can be extended in a non breaking way as well. We could make `sanitize` take a module name as an optional argument and simply default it to the current behavior (Wrap with `Ppx_deriving_runtime`).

Given the above, the current PR will work for more users of ppx_deriving. Especially those who might still want to user an older version of ppx_deriving while another plugin is still being updated.

**Comment by whitequark** [comment]
Well, what you did was a breaking change that necessitates a major bump. How this bump is handled is another story.

**Comment by rgrinberg** [comment]
I agree that ppx_deriving will require a major version bump. I'm just saying that it could be worthwhile to have a version of ppx_deriving_yojson that supports result and is 3.x and 4.x compatible

**Comment by rgrinberg** [comment]
I guess the real question is how do we fail about making the module name optional in sanitize. I'm not a big fan of piling up the optional args but here it seems to have tangible benefits.

**Comment by whitequark** [comment]
Sure, as long as it works.

**Comment by rgrinberg** [comment]
@whitequark to conclude this, I think this PR should be merged and a version of ppx_deriving_yojson should be released as is. I'll update ppx_deriving_yojson to use the newer api at a later time (will make a ticket so I don't forget)

**Comment by whitequark** [comment]
I'm unconvinced. Surely you could `opam pin` it for now, if you need it so quickly.

**Comment by rgrinberg** [comment]
Sure, I'm already pinning it so I'm in no rush. I just thought we agreed on the next steps to take. The two approaches are actually identical to me anyway, since I pretty much only use ppx_deriving and ppx_deriving_yojson there's no way the constraints could impact me. I just thought there was a to possibly minimize this pain for other users.

Ping me once you've decided how we should proceed.

**Comment by whitequark** [comment]
Just update this package to use the new API. I'll merge it and release it and ppx_deriving all at once when that's done.

**Comment by rgrinberg** [comment]
@whitequark ok I updated the PR to use the new api.

**Comment by whitequark** [comment]
Thanks!
